File: src/RequestPool.js
import { Constants } from "./Constants.js";

export class RequestPool {
    constructor(eventBus) {
        this.lock = new SimpleLock();
        this.pool = new Set();
        this.eventBus = eventBus;
    }

    add(request) {
        this.lock.waitForUnLock();
        this.pool.add(request);
    }

    /**
     * 将request通过EventBus发送
     */
    send() {
        this.lock.lock();
        this.eventBus.emit(Constants.SEND_FLOOR_REQUEST, [...this.pool]);
        this.unlock();
    }
}

File: src/main/java/com/example/demo/elevator/RequestPool.java
package com.example.demo.elevator;

import java.util.*;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

public class RequestPool {

    private final Lock lock = new ReentrantLock();
    private final Set<FloorRequest> pool = Collections.synchronizedSet(new HashSet<>());
    private final PositionManager pm;

    public RequestPool(PositionManager pm) {
        this.pm = pm;
    }

    /**
     * 向请求池中添加一个请求
     * @param request 楼层请求对象
     */
    public void add(FloorRequest request) {
        lock.lock();
        try {
            pool.add(request);
        } finally {
            lock.unlock();
        }
    }

    public void send() {
        lock.lock();
        try {
            // 通知pm处理请求
            pm.inserts(new ArrayList<>(pool).toArray(new FloorRequest[pool.size()]));

            // 发送后清空池子
            pool.clear();
        } finally {
            lock.unlock();
        }
    }
}